include/zephyr/kernel/obj_core.h

```diff
@@ -0,0 +1,283 @@
+#ifndef __KERNEL_OBJ_CORE_H__
+#define __KERNEL_OBJ_CORE_H__
+
+#include <zephyr/sys/dlist.h>
+
+/**
+ * @brief Macro to convert kernel object into pointer to object core
+ *
+ * This macro is used to convert a pointer to a kernel object (e.g. thread or
+ * memory block) into a pointer to its object core.
+ */
+#define K_OBJ_CORE(kobj)  (&((kobj)->obj_core))
+
+/* Known kernel object types */
+
+#define OBJ_TYPE_CONDVAR_ID    0x434f4e44   /* "COND" in ASCII */
+#define OBJ_TYPE_CPU_ID        0x4350555f   /* "CPU_" in ASCII */
+#define OBJ_TYPE_EVENT_ID      0x45564e54   /* "EVNT" in ASCII */
+#define OBJ_TYPE_FIFO_ID       0x4649464f   /* "FIFO" in ASCII */
+#define OBJ_TYPE_KERNEL_ID     0x4B524e4c   /* "KRNL" in ASCII */
+#define OBJ_TYPE_LIFO_ID       0x4c49464f   /* "LIFO" in ASCII */
+#define OBJ_TYPE_MEM_BLOCK_ID  0x4d424c4b   /* "MBLK" in ASCII */
+#define OBJ_TYPE_MBOX_ID       0x4d424f58   /* "MBOX" in ASCII */
+#define OBJ_TYPE_MEM_SLAB_ID   0x534c4142   /* "SLAB" in ASCII */
+#define OBJ_TYPE_MSGQ_ID       0x4d534751   /* "MSGQ" in ASCII */
+#define OBJ_TYPE_MUTEX_ID      0x4d555458   /* "MUTX" in ASCII */
+#define OBJ_TYPE_PIPE_ID       0x50495045   /* "PIPE" in ASCII */
+#define OBJ_TYPE_SEM_ID        0x53454d34   /* "SEM4" in ASCII */
+#define OBJ_TYPE_STACK_ID      0x5354434b   /* "STCK" in ASCII */
+#define OBJ_TYPE_THREAD_ID     0x54485244   /* "THRD" in ASCII */
+#define OBJ_TYPE_TIMER_ID      0x54494d52   /* "TIMR" in ASCII */
+
+struct k_obj_type;
+struct k_obj_core;
+
+#ifdef CONFIG_OBJ_CORE
+#define K_OBJ_CORE_INIT(_objp, _obj_type)   \
+	extern struct k_obj_type _obj_type; \
+	k_obj_core_init(_objp, &_obj_type)
+
+#define K_OBJ_CORE_LINK(objp) k_obj_core_link(objp)
+#else
+#define K_OBJ_CORE_INIT(objp, type)   do { } while(0)
+#define K_OBJ_CORE_LINK(objp)         do { } while(0)
+#endif
+/**
+ * Tools may use this list as an entry point to identify all registered
+ * object types and the object cores linked to them.
+ */
+extern sys_dlist_t obj_type_list;
+
+struct k_obj_type {
+	sys_dnode_t    node;   /* Node within list of object types */
+	sys_dlist_t    list;   /* List of objects of this object type */
+	uint32_t       id;     /* Unique type ID */
+	size_t         obj_core_offset;  /* Offset to obj_core field */
+};
+
+struct k_obj_core {
+	sys_dnode_t        node;   /* Object node within object type's list */
+	struct k_obj_type *type;   /* Object type to which object belongs */
+};
+
+/**
+ * @brief Initialize a specific object type
+ *
+ * Initializes a specific object type and links it into the object core
+ * framework.
+ *
+ * @param type Pointer to the object type to initialize
+ * @param id A means to identify the object type
+ * @param off Offset of object core within the structure
+ */
+extern void z_obj_type_init(struct k_obj_type *type, uint32_t id, size_t off);
```

Comment:
For completeness, might be best to return the struct k_obj_type pointer instead of void.  Any further use of the type is going to need that anyway.